scipy/linalg/_matfuncs.py

```diff
@@ -250,9 +270,141 @@ def expm(A):
 
     """
-    # Input checking and conversion is provided by sparse.linalg.expm().
-    import scipy.sparse.linalg
-    return scipy.sparse.linalg.expm(A)
+    a = np.asarray(A)
+    if a.size == 1 and a.ndim < 2:
+        return np.array([[np.exp(a.ravel()[0])]])
+
+    if a.ndim < 2:
+        raise LinAlgError('The input array must be at least two-dimensional')
+    if a.shape[-1] != a.shape[-2]:
+        raise LinAlgError('Last 2 dimensions of the array must be square')
+    n = a.shape[-1]
+    # Empty array
+    if min(*a.shape) == 0:
+        return np.empty_like(a)
+
+    # Scalar case
+    if a.shape[-2:] == (1, 1):
+        return np.exp(a)
+
+    if not np.issubdtype(a.dtype, np.inexact):
+        a = a.astype(float)
+    elif a.dtype == np.float16:
+        a = a.astype(np.float32)
+
+    # Explicit formula for 2x2 case, formula (2.2) in [1]
+    # without Kahan's method numerical instabilities can occur.
+    if a.shape[-2:] == (2, 2):
+        if a.ndim == 2:
+
+            p, r, s, t = a[0, 0], a[0, 1], a[1, 0], a[1, 1]
+            mu = csqrt((p-t)**2 + 4*r*s)/2
+            coshMu, sinchMu = np.cosh(mu), np.sinh(mu)/mu if mu != 0. else 1.
+            eA = np.empty([2, 2], dtype=mu.dtype)
+            eA[0, 0] = coshMu + 0.5*(p-t)*sinchMu
+            eA[0, 1] = r*sinchMu
+            eA[1, 0] = s*sinchMu
+            eA[1, 1] = coshMu - 0.5*(p-t)*sinchMu
+            eA *= np.exp((p + t)/2.)
+            if np.isrealobj(a):
+                return eA.real
+            return eA
+
+        else:
+            a1, a2, a3, a4 = (a[..., 0, 0],
+                              a[..., 0, 1],
+                              a[..., 1, 0],
+                              a[..., 1, 1])
+            mu = (a[..., 0, 0]-a[..., 1, 1])**2 + 4*a[..., 0, 1]*a[..., 1, 0]
+            mu = csqrt(mu)/2.
+            eApD2 = np.exp((a[..., 0, 0]+a[..., 1, 1])/2.)
+            AmD2 = (a1 - a4)/2.
+            coshMu = np.cosh(mu),
+            sinchMu = _sinch(mu)
+            eA = np.empty((a.shape), dtype=mu.dtype)
+            eA[..., 0, 0] = eApD2 * (coshMu + AmD2*sinchMu)
+            eA[..., 0, 1] = eApD2 * a2 * sinchMu
+            eA[..., 1, 0] = eApD2 * a3 * sinchMu
+            eA[..., 1, 1] = eApD2 * (coshMu - AmD2*sinchMu)
+            if np.isrealobj(a):
+                return eA.real
+            return eA
+
+    # Generic case with unspecified stacked dimensions.
+    # array to hold the result
+    n = a.shape[-1]
+    eA = np.empty(a.shape, dtype=a.dtype)
+    # working memory to hold intermediate arrays
+    Am = np.empty((5, n, n), dtype=a.dtype)
+
+    for ind in product(*[range(x) for x in a.shape[:-2]]):
+        aw = a[ind]
+
+        lu = bandwidth(aw)
+        if not any(lu):  # a is diagonal?
+            eA[ind] = np.diag(np.exp(np.diag(aw)))
+            continue
+
+        # # Generic/triangular case; copy the slice into scratch and send
+        Am[0, :, :] = aw
+        m, s = pick_pade_structure(Am)
```

Comment:
The other slices are going to be filled with higher powers of A during `pick_pade_structure`  then be sent to the actual machinery to compute the exponential. One way to shave some time off is to keep that scratch contiguous memory intact for reuse since it takes time to allocate and release memory.

Reply:
Ah, right, I didn't consider that the rest could be mutated by `pick_pade_structure`. I thought those would be set by `expm` a few lines later. Then it's OK, thanks.